fw/mmap_buffer.c

```diff
@@ -0,0 +1,303 @@
+ * plus TFW_MMAP_BUFFER_DATA_OFFSET; cpu_num is a number of CPU in a row. This
+ * allows determining which CPU's buffer should be mapped based on the offset.
+ */
+static int
+dev_file_mmap(struct file *filp, struct vm_area_struct *vma)
+{
+	TfwMmapBufferHolder *holder = filp->private_data;
+	TfwMmapBuffer *buf, *this_buf = *this_cpu_ptr(holder->buf);
+	unsigned long pfn, size, buf_size, buf_pages;
+	int cpu_num, cpu_id;
+
+	buf_size = TFW_MMAP_BUFFER_FULL_SIZE(this_buf->size);
+	size = vma->vm_end - vma->vm_start;
+	if (size > buf_size)
+		return -EINVAL;
+
+	buf_pages = buf_size / PAGE_SIZE;
+
+#define NTH_ONLINE_CPU(n) ({ \
+	int cpu, res = -1, i = 0; \
+	for_each_online_cpu(cpu) { \
+		if (i == n) { \
+			res = cpu; \
+			break; \
+		} \
+		++i; \
+	} \
+	res; \
+})
+
+	cpu_num = vma->vm_pgoff / buf_pages;
+	cpu_id = NTH_ONLINE_CPU(cpu_num);
+	if (cpu_id < 0)
+		return -EINVAL;
+
+	buf = *per_cpu_ptr(holder->buf, cpu_id);
+	pfn = page_to_pfn(virt_to_page(buf));
+
+	if (remap_pfn_range(vma, vma->vm_start, pfn, size, vma->vm_page_prot))
+		return -EAGAIN;
+
+	vma->vm_ops = &dev_vm_ops;
+	(void)dev_vm_ops;
+
+	if (size == buf_size)
+		atomic_set(&buf->is_ready, 1);
+
+	return 0;
+
+#undef NTH_ONLINE_CPU
+}
+
+static void dev_file_vm_close(struct vm_area_struct *vma)
+{
+	TfwMmapBufferHolder *holder = vma->vm_file->private_data;
+	TfwMmapBuffer *buf = *this_cpu_ptr(holder->buf);
+
+	atomic_set(&buf->is_ready, 0);
+}
+
+TfwMmapBufferHolder *
+tfw_mmap_buffer_create(const char *filename, unsigned int size)
+{
+	TfwMmapBufferHolder *holder;
+	unsigned int order;
+	int cpu;
+
+	if (size < TFW_MMAP_BUFFER_MIN_SIZE
+		|| size > TFW_MMAP_BUFFER_MAX_SIZE
+		|| !is_power_of_2(size))
+		return NULL;
+
+	if (filename && strlen(filename) >= TFW_MMAP_BUFFER_MAX_NAME_LEN - 1)
+		return NULL;
+
+	holder = kmalloc(sizeof(TfwMmapBufferHolder) +
+					 sizeof(struct page *) * num_online_cpus(),
+					 GFP_KERNEL);
+	if (!holder)
+		return NULL;
+
+	order = get_order(size);
+
+	holder->dev_major = -1;
+	holder->buf = __alloc_percpu_gfp(sizeof(TfwMmapBuffer *),
+									 sizeof(u64), GFP_KERNEL);
+	atomic_set(&holder->dev_is_opened, 0);
+	atomic_set(&holder->is_freeing, 0);
+
+	for_each_online_cpu(cpu) {
+		TfwMmapBuffer *buf, **bufp;
+
+		holder->pg[cpu] = alloc_pages_node(cpu_to_node(cpu), GFP_KERNEL, order);
+		if (holder->pg[cpu] == NULL)
+			goto err;
+
+		buf = (TfwMmapBuffer *)page_address(holder->pg[cpu]);
+		buf->size = TFW_MMAP_BUFFER_DATA_SIZE(size);
+		buf->head = 0;
+		buf->tail = 0;
+		buf->cpu = cpu;
+		bufp = per_cpu_ptr(holder->buf, cpu);
+		*bufp = buf;
+		atomic_set(&buf->is_ready, 0);
+	}
+
+	if (filename) { /* do not create the file in unit tests */
+		holder->dev_major = register_chrdev(0, filename, &dev_fops);
+		if (holder->dev_major < 0) {
+			T_WARN("Registering char device failed for %s\n", filename);
+			goto err;
+		}
+
+		holder->dev_class = class_create(THIS_MODULE, filename);
+		device_create(holder->dev_class, NULL,
```

Comment:
device_create can also fails.so we should handle this case

Reply:
Fixed.